### PR TITLE
Add i18next.

### PR DESCRIPTION
**Pull request by NorwegianKiwi**
Add i18next to be able to have captions in several languages. For now, english (en-US) and norwegian (nb-NO), and only for intro texts (intro.js). There are several things that are not translated yet, including the html in the jade layouts. See comments in i18n.js.

Hopefully, if all the text in codeclub_lesson_builder can be included using i18n, perhaps the nb-NO-branch can be merged with master and then deleted.

### CODE REVIEW

**Comment by NorwegianKiwi** [comment]
Note that the default language is now set to norwegian (see option "lng" in i18n.js). Since this is the master branch, it should perhaps be changed to english?

File: gulpfile.js [hunk 1]
```diff
@@ -94,7 +103,7 @@ gulp.task('browserify', function() {
   .transform("babelify", {presets: ["es2015"]})
   .bundle()
   .pipe(source('script.min.js'))
-  .pipe(streamify(uglify()))
+  //.pipe(streamify(uglify()))
```
Comment (arve0) [inline gulpfile.js]: debug?

Comment (arve0) [inline gulpfile.js]: Should add a source map..been on my mind.

File: locales/en-US/translation.json [hunk 2]
```diff
@@ -0,0 +1,29 @@
+{
```
Comment (arve0) [inline locales/en-US/translation.json]: I guess this file is required by i18next. Would it be possible to use a `.js` file which uses `module.export`? Then one can add comments.

**Comment by arve0** [comment]
Nice work!

> Hopefully, if all the text in codeclub_lesson_builder can be included using i18n, perhaps the nb-NO-branch can be merged with master and then deleted.

Then we need to put the configuration in the lesson repo? #155 may make this easier.

>  Since this is the master branch, it should perhaps be changed to english?

Yes.

Would you like this to be merged now, before moving the layouts over to i18next?

File: gulpfile.js (hunk 1 above)
Comment (NorwegianKiwi) [inline gulpfile.js]: Oops, yes, didn't mean to check in that. Will fix.

Sourcemap would be nice, can perhaps see what I can do later (could make an issue).

File: locales/en-US/translation.json (hunk 2 above)
Comment (NorwegianKiwi) [inline locales/en-US/translation.json]: I'm not sure if having the translations in .js files is possible (I could look some more into it), but what do you think about something like https://github.com/sindresorhus/gulp-strip-json-comments ?
Since the json-files need to be copied to the build folder anyway, they might as well have their comments stripped on the way.

**Comment by NorwegianKiwi** [comment]
Why do you say that the config would need to be in the lesson repo? Were you thinking that the lessons also should be in both languages?

And you are probably right, I should do the layouts before merging in the changes. I was thinking of using https://github.com/i18next/jquery-i18next for this.

File: locales/en-US/translation.json (hunk 2 above)
Comment (arve0) [inline locales/en-US/translation.json]: If .js is not possible, stripping comments at build is a nice supplement. I'm thinking about the clarity of the keys, comments might be useful.

**Comment by arve0** [comment]
> config ...

I'm thinking that you need to specify the language somewhere. But that might be better to have in the front end? Might be what you have thinking all along?

Instead of jquery, what about defining i18next in the layout, then build both languages?
https://github.com/arve0/codeclub_lesson_builder/blob/master/build.js#L50

**Comment by NorwegianKiwi** [comment]
The code that I now added to this pull request will build the static files once (currently english, but by changing "lng" in build-with-i18n.js it can build a different language). The language of the rest can be changed in the client. It still means that a manual change + rebuild needs to be done if one wants to change the language of the website, but it is a step on the way.

I want to be able to change the language on the client, on the fly. I have tried to include the plugin metalsmith-i18n, but it seems like it is using an older version of i18n and did not work to my satisfaction. (I might of course have misunderstood how metalsmith-i18n works and done it suboptimally, but...)

I think, at this point, the best option is NOT to build all the html-files (via metalsmith and jade) twice, but rather find a way to only have one copy of each (since it is hard to make sure all the links are correct for each language, and also it seems unnecessary to have two of every html-file).

I have two ideas so far. One is to first create tags with IDs for all the text that needs translation, and then make metalsmith "hardcode" all the translations into a script in the head of layout.jade, and have the same script change the tags with the given IDs. Then one can change the language with e.g. "?lng=nb-NO" (and perhaps set it in a session variable, or even as a cookie) --- and perhaps add a couple of buttons/links/flags to change the language as well.

The other idea is to change the front to use React. That way React can rerender the page with the new language whenever it is changed by the client.

**Comment by NorwegianKiwi** [comment]
Hm, I see that the build failed. Looking into that...

**Comment by NorwegianKiwi** [comment]
I have now added the captions in the html-files (created by metalsmith and jade) by using the attribute "data-i18n" in the tags that need captions (see i18n.js for details on how to include translations for both the innerHTML and attributes). This way we don't need to compile two sets of html-files, and one can easily switch on the fly. The language is stored in a cookie so that it remembers it the next time you go to the webpage.

I have also added flags in the topmenu to quickly switch between english and norwegian. These can easily be turned on and off in config.js (showFlags = true og false), and the rebuild.

I have translated most of the captions, but I think there are a couple that remain. These should be easy to add later.

**Comment by NorwegianKiwi** [comment]
I think this pull-request is ready to be merged. Hopefully we can get rid of the nb-NO branch now!

**Comment by arve0** [comment]
Thanks, I see you have put in a lot of work! As this change is quite comprehensive, I'll need some time to review it.

A few initial thoughts:

  1. We should document this change.
  2. Is initial "build" possible in the templates?
  3. Related to point above, is there any SEO we should think about? In example [`title`](https://github.com/NorwegianKiwi/codeclub_lesson_builder/blob/i18n/layouts/layout.jade#L5) is a bit unfortunate, as "kodeklubben oppgaver" would be a top search term. `kodeklubben` is in the url, so this might not change search results. But it will change the search result title, which is a part of the authenticity.

**Comment by NorwegianKiwi** [comment]
1. Completely agree! Do you have any thoughts on where? In the readme.md?
2. Not sure what you mean with this question. Do you mean that there should be default (hardcoded) texts in a language given by the build config? (Right now, the texts are empty until the page is loaded, and then filled in). If this is the question, then it is not possible right now, since I removed the use of i18n in the build-script, but if I reintroduce it from an earlier commit, then it should be doable.
3. I don't really know much about tuning SEO, and not sure what you mean when you say that "title" is unfortunate. The document.title should be the same as before, it is only if it is missing from the metalsmith templates (in oppgaver/src) that the title in the translation.json is used.

**Comment by arve0** [comment]
I've started reviewing. A bit hard to follow the flow, but getting there. Hope to get it done tonight, if not I will finish it in the week to come.

1. I'm thinking of writing some few words in a separate file, like STRUCTURE.md. I'll take a stab at it when finished reviewing this.
2. What I'm thinking of is adding `i18n_t('key')` to layouts. Then in client side, only do `i18n_t('key')` if language is changed explicit (clicking the flag). Example:

  ```js
  // from
  a.download(href=link.replace('.html', '.pdf') data-i18n="title=layout.downloadPdf")
  // to
  a.download(href=link.replace('.html', '.pdf') data-i18n="title=layout.downloadPdf" title=i18n_t("layout.downloadPdf"))
  ```

3. Search engines do not run JS. **Old**: 
  old
  vs **new**:
  new

PS: Thanks for the the work this far.

**Comment by NorwegianKiwi** [comment]
A few comments:

2. This can relatively easily be done by reintroducing the code I had in commit d5a3623fd2cdbb66c8dc6014ed911b46a91c243c.

3. Ah, my bad, I thought that "pipe" was a logical "or", which is why I removed it. Easily reinserted (especially if the code mentioned above is reintroduced).

Do you want med to do this? It shouldn't take too much time.

Also, feel free to bug me on Facebook if you have any questions! :)

**Comment by arve0** [comment]
> Do you want med to do this? It shouldn't take too much time.

Sure, though I think it will be a good practise for me :wink:

**Comment by arve0** [comment]
> Personally I must admit that I am more used to 4-space indent, but it seems like 2-space has been used a lot already, so seems fair to keep it that way. Btw, reason I didn't fix the indents in the last PR was that I wanted to make as few changes as possible. But good to have it properly indented.
One small suggestion: What do you think about adding the option " load:'currentOnly' " to i18n? If not, then a lot of warnings pop up in the console (due to the whitelist) when using debug:true. I am going to merge the PR anyway, we can always add this afterwards.

Funfact, in Github's top repositories, 60% are 2 spaces, 27% are 4 spaces. But more seriously, this should be fixed by a tool. Manually indenting is sooo last year :wink: 

> I also noticed that layout.jade has no hardcoded translations in it yet. Also, the title is still "Oppgaver" and not "Oppgaver | Kodeklubben". Do you want me to fix any of these after the merge before you do the merge of the original PR?

I will add this.

**Comment by NorwegianKiwi** [comment]
Ref "force a language": I suggest to set "showFlags" to false (in config.js), thus hiding the flags. It will still be possible to change language using ?lng=en-US, but it won't be available by clicking. But this way, we have both translations there all the time, and don't need two separate branches for the languages.

In order to make sure it always uses norwegian, we could add lng:'nb-NO' to the i18n.init() options, thus choosing this automatically (unless it was set to english previously and stored in localstorage).

**Comment by NorwegianKiwi** [comment]
60% is 2-spaces? Interesting! Also, what are the remaining 13%? 3 spaces? Tabs?

**Comment by arve0** [comment]
> 60% is 2-spaces? Interesting! Also, what are the remaining 13%? 3 spaces? Tabs?
 Show all checks

In javascript, yes. Remaining are tabs. Another funfact, 4 spaces is 92% in python, but Ruby is 100% 2 spaces.

> In order to make sure it always uses norwegian, we could add lng:'nb-NO' to the i18n.init() options, thus choosing this automatically (unless it was set to english previously and stored in localstorage).

I think specifying with `lng` would be sufficient.

**Comment by arve0** [comment]
Thanks @NorwegianKiwi! nb-NO branch can be deleted after the layouts/styles have been merged back into master.
